okhttp/src/jvmMain/kotlin/okhttp3/internal/ws/RealWebSocket.kt

```diff
@@ -560,40 +571,66 @@ class RealWebSocket(
     }
 
     if (failedPing != -1) {
-      failWebSocket(SocketTimeoutException("sent ping but didn't receive pong within " +
-          "${pingIntervalMillis}ms (after ${failedPing - 1} successful ping/pongs)"), null)
+      failWebSocket(
+        e = SocketTimeoutException(
+          "sent ping but didn't receive pong within " +
+            "${pingIntervalMillis}ms (after ${failedPing - 1} successful ping/pongs)"
+        ),
+        isWriter = true,
+      )
       return
     }
 
     try {
       writer.writePing(ByteString.EMPTY)
     } catch (e: IOException) {
-      failWebSocket(e, null)
+      failWebSocket(e = e, isWriter = true)
     }
   }
 
-  fun failWebSocket(e: Exception, response: Response?) {
+  fun failWebSocket(
+    e: Exception,
+    response: Response? = null,
+    isWriter: Boolean = false,
+  ) {
+    val streamsToCancel: Streams?
     val streamsToClose: Streams?
-    val readerToClose: WebSocketReader?
     val writerToClose: WebSocketWriter?
     synchronized(this) {
       if (failed) return // Already failed.
       failed = true
-      streamsToClose = this.streams
-      this.streams = null
-      readerToClose = this.reader
-      this.reader = null
+
+      streamsToCancel = this.streams
+
       writerToClose = this.writer
       this.writer = null
+
+      streamsToClose = when {
+        writerToClose != null && reader == null -> this.streams
+        else -> null
+      }
+
+      if (!isWriter && writerToClose != null) {
+        // If the caller isn't the writer thread, get that thread to close the writer.
+        taskQueue.execute("$name writer close", cancelable = false) {
+          writerToClose.closeQuietly()
+          streamsToClose?.closeQuietly()
+        }
+      }
+
       taskQueue.shutdown()
     }
 
     try {
       listener.onFailure(this, e, response)
     } finally {
-      streamsToClose?.closeQuietly()
-      readerToClose?.closeQuietly()
-      writerToClose?.closeQuietly()
+      streamsToCancel?.cancel()
+
+      // If the caller is the writer thread, close it on this thread.
```

Comment:
Does this logic get simpler with some `taskQueue.executeOrImmediate()` method that is inlined if already on the task queue? It's the same code as line 616, we just don't want the overhead of execute.

But maybe that's just a flourish?

Reply:
One gotcha is the synchronized statement. I wanna decide what to do in the synchronized block, but do it outside of that block. So the simple abstraction is difficult!